ambry-server/src/main/java/com.github.ambry.server/AmbryServer.java

```diff
@@ -65,28 +67,33 @@
   private ReplicationManager replicationManager = null;
   private Logger logger = LoggerFactory.getLogger(getClass());
   private final VerifiableProperties properties;
-  private final ClusterMap clusterMap;
+  private final ClusterAgentsFactory clusterAgentsFactory;
+  private ClusterMap clusterMap;
+  private ClusterParticipant clusterParticipant;
   private MetricRegistry registry = null;
   private JmxReporter reporter = null;
   private ConnectionPool connectionPool = null;
   private final NotificationSystem notificationSystem;
   private ServerMetrics metrics = null;
   private Time time;
 
-  public AmbryServer(VerifiableProperties properties, ClusterMap clusterMap, Time time) throws IOException {
-    this(properties, clusterMap, new LoggingNotificationSystem(), time);
+  AmbryServer(VerifiableProperties properties, ClusterAgentsFactory clusterAgentsFactory, Time time)
+      throws IOException {
+    this(properties, clusterAgentsFactory, new LoggingNotificationSystem(), time);
   }
 
-  public AmbryServer(VerifiableProperties properties, ClusterMap clusterMap, NotificationSystem notificationSystem,
-      Time time) throws IOException {
+  AmbryServer(VerifiableProperties properties, ClusterAgentsFactory clusterAgentsFactory,
+      NotificationSystem notificationSystem, Time time) throws IOException {
     this.properties = properties;
-    this.clusterMap = clusterMap;
+    this.clusterAgentsFactory = clusterAgentsFactory;
     this.notificationSystem = notificationSystem;
     this.time = time;
   }
 
-  public void startup() throws InstantiationException {
+  void startup() throws InstantiationException {
     try {
+      clusterMap = clusterAgentsFactory.getClusterMap();
```

Comment:
Probably wrote it this way because `startup()` seems to do most of the heavy-lifting. And now these involve connecting to a service over the network. `AmbryServer` constructor in fact does not throw (contrary to what the signature says).